### PR TITLE
Add automatic conversion for other number bases

### LINKED ISSUE #1854: Add support for hexadecimal numbers without conversion
can someone have a look at these blocks? I can't see any error

blocks (1)

i get `0 - 0 - 0`

### PR DESCRIPTION
**Pull request by BeksOmega**
### Resolves

Closes #1854 

### Description

Makes it so that the number field accepts hex, binary, and octal
formats in addition to the binary format. Numbers are automatically
converted to decimal upon code generation.

#### Decimal
UI:
Formats1
Generated Code:
FromatsS-1

#### Hex
UI:
Formats2
Formats3
Generated Code:
FormatS-2

#### Octal
UI:
Formats4
Generated Code:
FormatS-3

#### Binary
UI:
Formats5
Generated Code:
FormatsS-4

### Additional Information

This is just my interpretation of how the issue should be fixed. So if you want anything to work differently I'm totally up for some reworking!

### CODE REVIEW

**Comment by AppInventorWorkerBee** [comment]
Can one of the admins verify this patch?

**Comment by ewpatton** [comment]
One of the benefits of a visual language is that you can show things and make them discoverable. What I don't like about using the C-like prefixes is that unless you already know about them, it's really hard to figure out that they would even exist. I think what would be more useful is a version of the number block that had a dropdown field that controlled the base in place of having a prefix (you could even have switching the dropdown reinterpret the numeric value in the new base, e.g., if you have hexadecimal ff and switch to decimal it would change the text value to be 255). Then, this block is visible in the drawer and people can see it.

**Comment by ewpatton** [comment]
I think we can keep this implementation as well for more advanced users, but I'm not sure that it's the right approach for beginners.

**Comment by ewpatton** [comment]
After further consideration, I think my first suggestion should be submitted as a separate PR (needs to be against `ucr` due to new block).

**Comment by ewpatton** [comment]
@AppInventorWorkerBee ok to test

**Comment by BeksOmega** [comment]
@ewpatton Documentation has been added, and critiques are welcome :D 

> a version of the number block that had a dropdown field that controlled the base in place of having a prefix (you could even have switching the dropdown reinterpret the numeric value in the new base, e.g., if you have hexadecimal ff and switch to decimal it would change the text value to be 255)

Should be totally doable! And it sounds like a fun project :D 

My one question is what would you like it to output? Would you want it to be a decimal number (like the above) or a string (e.g. ff)?

If I'm understanding your type system correctly: If you want it to output a decimal number it wouldn't work with the `convert number` block. But if you want it to output a string it wouldn't work with the other number blocks.

**Comment by ewpatton** [comment]
@BeksOmega As a first pass the return values for the new block should still be numbers. Numbers always render in base 10, so that's a bit unfortunate, but it will work. If you're feeling really ambitious, you could implement a subclass of gnu.math.IntNum that also stores the rendering base with a toString method that renders in the base accordingly.

File: appinventor/docs/markdown/reference/blocks/math.md [hunk 1]
```diff
@@ -79,7 +79,12 @@ convert radians to degrees, convert degrees to radians
 
 ![](images/math/number.png)
 
-Can be used as any positive or negative number (decimals included). Clicking on the "0" in the block will allow you to change the number.
+Can be used as any positive or negative number. Clicking on the "0" in the block will allow you to change the number.
+
+The block supports normal base-10 numbers (for example: 2, 12, and 2.12), as well as C-like prefixes for other number bases. It supports:
```
Comment (ewpatton) [inline appinventor/docs/markdown/reference/blocks/math.md]: Maybe in parentheses after the base value show the decimal value, e.g.,

> * Base-2 (binary) numbers, e.g., 0b10 (decimal 2)
> * Base-8 (octal) numbers, e.g., 0o14 (decimal 12)
> * Base-16 (hexadecimal) numbers, e.g., 0xd4 (decimal 212)

**Comment by BeksOmega** [comment]
@ewpatton I think you're right, adding the decimal conversion makes it clearer. Updated :D 

Also - creating a new dropdown number block is in my task management system, but I think I'm going to work on finishing up some of my other projects (e.g. collapsing) first. I'm always up for re-prioritizing though, so if you want me to switch something around just shout.

**Pull request merged**